proposals/cleanup.md

```diff
@@ -0,0 +1,256 @@
+# Meta
+
+- Name: Clean-up
+- Start Date: 2022-07-30
+- Author(s): chipzoller
+- Supersedes: N/A
+
+# Table of Contents
+
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+- [CRD Changes (OPTIONAL)](#crd-changes-optional)
+
+# Overview
+
+A clean-up ability of Kyverno allowing it to clean-up (delete) resources in a Kubernetes cluster based upon certain criteria.
+
+# Definitions
+
+- TTL: Time-to-live. The amount of time a resource may exist until it is cleaned up.
+
+# Motivation
+
+The initial idea for this feature came from issue [3483](https://github.com/kyverno/kyverno/issues/3483) authored by user luisdavim.
+
+As long as there has been Kubernetes, there has been the inevitability that some resources--most especially those created directly by human users--are left derelict at some point without a good way to identify what and where they are. As Kubernetes continues to expand with its myriad of possible custom resources, this problem grows over time. The ultimate problems this poses are few: 1) Kubernetes' key-value store, which has fairly limited storage capacity, may become overwhelmed by resources which should be deleted; and 2) these resources, in the case of Pods, may result in consumption of excess resources driving up cost and reducing workload consolidation abilities. The solution to these problems is to identify the derelict resources and remove (i.e., delete) them.
+
+There are two main use cases associated with this.
+
+1. Explicit, per-resource clean-up declaration which may be set by a user either upon creation or retroactively.
+
+2. Implicit, bulk clean-up of resources on a scheduled basis which meet a certain criteria.
+
+
+# Proposal
+
+In this proposal, Kyverno may function as a clean-up controller allowing it to identify and scavenge resources. There are two proposed capabilities which would be used to effect a clean-up.
+
+1. Use of two specific annotations which may be set on any resource:
```

Comment:
We will need labels to register the webhookconfigurations to receive admission requests for matching resources. Note that the label value does not accept `:`.

Reply:
Yes, "labels" = "annotations" in this proposal. As discussed, we probably need to use labels for the `ttl` and `expires` functionalities so a separate rule in the ValidatingWebhookConfiguration can be created which only matches on those two specific labels.